src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -471,19 +489,114 @@ void enforceMemberNotNull(SyntaxNode? syntaxOpt, LocalState state)
                 var method = _symbol as MethodSymbol;
                 if (method is object)
                 {
-                    do
+                    if (method.IsConstructor() && _useConstructorExitWarnings)
                     {
-                        foreach (var memberName in method.NotNullMembers)
+                        var thisSlot = 0;
+                        if (method.RequiresInstanceReceiver)
                         {
-                            enforceMemberNotNullOnMember(syntaxOpt, state, method, memberName);
+                            method.TryGetThisParameter(out var thisParameter);
+                            Debug.Assert(thisParameter is object);
+                            thisSlot = GetOrCreateSlot(thisParameter);
                         }
+                        // https://github.com/dotnet/roslyn/issues/46718: give diagnostics on return points, not constructor signature
+                        var exitLocation = /*syntaxOpt?.Location ??*/ (method.DeclaringSyntaxReferences.IsEmpty ? null : method.Locations.FirstOrDefault());
+                        foreach (var member in method.ContainingType.GetMembersUnordered())
+                        {
+                            checkMemberStateOnConstructorExit(method, member, state, thisSlot, exitLocation);
+                        }
+                    }
+                    else
+                    {
+                        do
+                        {
+                            foreach (var memberName in method.NotNullMembers)
+                            {
+                                enforceMemberNotNullOnMember(syntaxOpt, state, method, memberName);
+                            }
 
-                        method = method.OverriddenMethod;
+                            method = method.OverriddenMethod;
+                        }
+                        while (method != null);
                     }
-                    while (method != null);
                 }
             }
 
+            void checkMemberStateOnConstructorExit(MethodSymbol constructor, Symbol member, LocalState state, int thisSlot, Location? exitLocation)
+            {
+                var isStatic = !constructor.RequiresInstanceReceiver();
+                if (member.IsStatic != isStatic)
+                {
+                    return;
+                }
+
+                // This is not required for correctness, but in the case where the member has
+                // an initializer, we know we've assigned to the member and
+                // have given any applicable warnings about a bad value going in.
+                // Therefore we skip this check when the member has an initializer to reduce noise.
```

Comment:
If it is set in the constructor body, we know we will give a warning at the point that a bad value goes in.

Reply:
It feels like this is inconsistent about warnings: if I assign a bad value in the constructor, I get a warning where that was assigned. However, depending on whether the property had an initializer or not, I may get an additional warning about the state at the end of the constructor body. I think we should be consistent here: either we always warn at the end of a body (the easier to implement approach), or we only warn if we didn't warn about a bad value going in (the harder approach). Doing otherwise will feel like a bug from the user's perspective IMO.

---
In reply to: [469259612](https://github.com/dotnet/roslyn/pull/46441#discussion_r469259612) [](ancestors = 469259612)

Reply:
The main fallout of this is a lot of test churn, so I'd like to do it in a follow-up if that's acceptable.

Reply:
Probably at the same time that I move the location of uninitialized field warnings. #46718

Reply:
That's fine with me.

---
In reply to: [470202737](https://github.com/dotnet/roslyn/pull/46441#discussion_r470202737) [](ancestors = 470202737)